cdap-app-fabric/src/main/java/co/cask/cdap/common/twill/TwillAppLifecycleEventHandler.java

```diff
@@ -0,0 +1,236 @@
+  private AtomicBoolean launchedContainer;
+  private AtomicBoolean stoppedContainer;
+
+  /**
+   * Constructs an instance of TwillAppLifecycleEventHandler that abort the application if some runnable has no
+   * containers and a program state writer to respond to Twill application lifecycle events
+   *
+   * @param abortTime Time in milliseconds to pass before aborting the application if no container is given to
+   *                  a runnable.
+   * @param options the program options, if the event handler was attached to monitor program states
+   */
+  public TwillAppLifecycleEventHandler(long abortTime, ProgramOptions options) {
+    this(abortTime, false, options);
+  }
+
+  /**
+   * Constructs an instance of TwillAppLifecycleEventHandler that abort the application if some runnable has not enough
+   * containers.
+   * @param abortTime Time in milliseconds to pass before aborting the application if no container is given to
+   *                  a runnable.
+   * @param abortIfNotFull If {@code true}, it will abort the application if any runnable doesn't meet the expected
+   *                       number of instances.
+   * @param options the program options, if the event handler was attached to monitor program states
+   */
+  public TwillAppLifecycleEventHandler(long abortTime, boolean abortIfNotFull, ProgramOptions options) {
+    super(abortTime, abortIfNotFull);
+    this.programRunId = options.getProgramId().run(ProgramRunners.getRunId(options));
+  }
+
+  @Override
+  protected Map<String, String> getConfigs() {
+    Map<String, String> configs = super.getConfigs();
+    configs.put("programRunId", GSON.toJson(programRunId));
+    return configs;
+  }
+
+  @Override
+  public void initialize(EventHandlerContext context) {
+    super.initialize(context);
+
+    this.twillRunId = context.getRunId();
+    this.programRunId = GSON.fromJson(context.getSpecification().getConfigs().get("programRunId"), ProgramRunId.class);
+
+    // Fetch cConf and hConf from resources jar
+    File cConfFile = new File("resources.jar/resources/" + CDAP_CONF_FILE_NAME);
+    File hConfFile = new File("resources.jar/resources/" + HADOOP_CONF_FILE_NAME);
+
+    if (cConfFile.exists() && hConfFile.exists()) {
+      CConfiguration cConf = CConfiguration.create();
+      cConf.clear();
+
+      Configuration hConf = new Configuration();
+      hConf.clear();
+
+      try {
+        cConf.addResource(cConfFile.toURI().toURL());
+        hConf.addResource(hConfFile.toURI().toURL());
+
+        // Create the injector to inject a program state writer
+        Injector injector = Guice.createInjector(
+          new ConfigModule(cConf, hConf),
+          new ZKClientModule(),
+          new KafkaClientModule(),
+          new DiscoveryRuntimeModule().getDistributedModules(),
+          new AbstractModule() {
+            @Override
+            protected void configure() {
+              bind(RuntimeStore.class).to(RemoteRuntimeStore.class);
```

Comment:
Yes, the DirectStoreProgramStateWriter needs that binding as the type is a RuntimeStore